proposals/NNNN-unlock-existentials-with-fixed-associated-types.md

```diff
@@ -0,0 +1,237 @@
+
+```swift
+protocol User: Identifiable 
+    where ID == String {
+    
+    var username: String { get }
+    var displayName: String { get }
+}
+
+extension User {
+    var id: String {
+        username
+    }
+}
+```
+
+Many would rightfully assume that `User` could be used as a Type, since we specify that the `ID` asscoated type inherited from [`Identifiable`](https://github.com/apple/swift-evolution/blob/master/proposals/0261-identifiable.md) is `String`. 
+
+```swift
+let myUser: User
+// ❌ `User` is not
+// usable as a Type
+```
+> **_NOTE:_** One may point out that `myUser` could be bound to `some User`, which would have a similar effects. However, [opaque result types](https://github.com/apple/swift-evolution/blob/master/proposals/0244-opaque-result-types.md) are not Existentials. That is, the former allows for hiding type information from the user while the compiler internally retains the underlying type. On the contrary, the latter properly erases type information allowing storage of different types - that conform to a given protocol. What that means, is that should `myUser` be bound to `some User` then initialization would have to be performed at the declaration-site and mutation would be prohibited.
+
+This is a great inconvenience with not so elegant workarounds:
+
+```swift
+protocol AnyUser { ... } 
+// No `Identifiable` conformance
+
+let myUser: AnyUser
+// ✅
+
+
+protocol User: AnyUser, Identifiable 
+    where ID == String { 
+    
+    var id: String {
+        username
+    }
+}
+
+let myOtherUser: User 
+// ❌ Still an error as expected
+```
+
+These workarounds, besides producing boilerplate and confusing code, also produce confusing API designs leaving clients to wonder which of the two protocols should be used. In other cases, API authors may decide that adding `Identifiable` isn’t worth the added complexity and leave the `User` protocol with no such conformance whatsoever. 
+
+As a result, a straightforward conformance to simple - yet quite useful - protocols such as `Identifiable` and others becomes a hassle to implement.
+
+
+## Proposed solution
+
+We propose to simply allow User-like protocols to be used as Types. Thus, making more natural code possible - which fits a goal of Swift of building ["expressive and elegant APIs"](https://forums.swift.org/t/on-the-road-to-swift-6/32862). Furthermore, library authors could start adding useful protocol conformances to their protocols - a task that's currently prohibitively complex.
+
+
+## Detailed design
+
+### Which Protocols would NOT be Usable as Types 
+
+Now, a protocol is considered unusable as a Type when it:
+
+1. Includes at least one associated type or `Self` requirement - which may have been inherited - and
+2. if at least one inherited associated type requirement - if there is any - is _not_ specified.
```

Comment:
The proposal is relaxing a restriction. Could we define this in terms of the protocols that *will* be supported? There is no need to have separate "Protocols that would (NOT) be Usable as Types" sections then and we can simply list the examples.

Reply:
I too was skeptical about this part, I'll take a look.

Reply:
I changed the definition to:

> Now, a protocol is considered usable as a Type when it:
>
> 1. Does _not_ include `Self` requirements - and
> 2. if it has no associated type requirements or if all its associated types are fixed.

Do you think "fixed" is the right work here or should we go for something like "known"?

Reply:
> Do you think "fixed" is the right work here or should we go for something like "known"?

Yes, I do think we should generally switch to "known", because a same-type constraint is not the only means of knowing the implementation of an associated type requirement.